### PR TITLE
Fetch pushToken from Expo every week

### PR DESCRIPTION
**Pull request by andyguest**
Let's be more defensive about this, even though according to Expo themselves, push tokens should remain valid for as long as the app remains installed.  See: https://forums.expo.io/t/exponentpushtoken-when-does-it-really-change/3611/2

Let's also clear the push token if someone asks us to clear data.

N.B. UNTESTED: want to hear any comments before verifying (perhaps in Stage).

# Description

_Description of the feature or fix with link to the issue if existing_

## On what platform have you tested the change?

- [ ] Android - Emulator
- [ ] Android - Physical device
- [ ] iOS - Emulator
- [ ] iOS - Physical device

## Screenshots

_Please attach screenshots showing the change if relevant_

## Checklist

- [ ] I have updated mockServer
- [ ] I've added analytics as relevent
- [ ] I have run `npm test`
- [ ] I have run `npm run test:i18n`

## Out of scope and potential follow-up

_Is there any related changes that you plan to do in a follow-up PR or voluntarily excluded from the scope?_

### CODE REVIEW

File: src/core/push-notifications/PushNotificationService.ts [hunk 1]
```diff
@@ -70,51 +70,48 @@ export default class PushNotificationService {
...
+    } catch (error) {
+      // Fail silently
     }
   }
 
   static async openSettings() {
     switch (Platform.OS) {
       case 'ios':
-        Linking.openURL('app-settings:');
+        await Linking.openURL('app-settings:');
```
Comment (selvach) [inline src/core/push-notifications/PushNotificationService.ts:111]: Remove this await. It causes issues.

File: src/core/push-notifications/PushNotificationService.ts [hunk 2]
```diff
@@ -70,51 +70,48 @@ export default class PushNotificationService {
...
         break;
       case 'android':
-        IntentLauncher.startActivityAsync(IntentLauncher.ACTION_NOTIFICATION_SETTINGS);
+        await IntentLauncher.startActivityAsync(IntentLauncher.ACTION_NOTIFICATION_SETTINGS);
```
Comment (selvach) [inline src/core/push-notifications/PushNotificationService.ts:114]: Remove this await. It causes issues.

File: src/core/push-notifications/PushNotificationService.ts [hunk 3]
```diff
@@ -70,51 +70,48 @@ export default class PushNotificationService {
...
-        Analytics.track(events.NOTIFICATION_ENABLED);
-      } catch (error) {
-        // Fail silently
+      if (notify_backend && pushToken) {
+        // Send to our backend first and save locally only once successful.
+        await this.apiClient.updatePushToken(pushToken);
+        await this.savePushToken(pushToken);
       }
+    } catch (error) {
+      // Fail silently
```
Comment (selvach) [inline src/core/push-notifications/PushNotificationService.ts:104]: Probably want to do something here...

**Review by selvach — CHANGES_REQUESTED** [review]
Could you just merge in these unit tests on the `tokenNeedsRefreshing()` function, please!

https://github.com/zoe/covid-tracker-react-native/compare/selvach/force_new_token

File: src/core/push-notifications/PushNotificationService.ts (hunk 3 above)
Comment (bionara) [reply src/core/push-notifications/PushNotificationService.ts:104]: Fail Noisily @andyguest :)

**Review by bionara — DISMISSED** [review]
Code looks good once Somesh's feedback is implemented

File: src/core/push-notifications/PushNotificationService.ts (hunk 3 above)
Comment (andyguest) [reply src/core/push-notifications/PushNotificationService.ts:104]: I'm maintaining the old error handling code here 😁 , but note that the silent failure still triggers a retry on next app opening.  Not sure if we can do much better than that?  Don't think we want to interrupt users with a pop-up for example, but maybe you'd prefer we did?

**Review by selvach — APPROVED** [review]

**Pull request merged**